### DISCUSSION TITLE
Start of new SPI transaction triggered by counter/timer interrupt.

### DISCUSSION THREAD

**Discussion by cz8pru** [zephyrproject-rtos/zephyr#40406] [Q&A]
https://github.com/zephyrproject-rtos/zephyr/blob/5a39a73c4e6dfc21055ae5bb8e82eecd269a6750/drivers/spi/spi_context.h#L82

I have issue with my application, where I would like to start (by using DMA) SPI transaction from interrupt (precious timing us).
When I wanted to run it, program always finished here ( `z_impl_k_sem_take()` ):

`__ASSERT(((arch_is_in_isr() == false) || K_TIMEOUT_EQ(timeout, K_NO_WAIT)), "");`

Due to the fact that "The kernel does allow an ISR to take a semaphore, however the ISR must not attempt to wait if the semaphore is unavailable" this SPI driver cannot work.

My workaround is in spi_context_lock() to change `//k_sem_take(&ctx->lock, K_FOREVER);` to `k_sem_take(&ctx->lock, K_NO_WAIT);`

But from system point of view, we have to get option to use it as a blocking or not.....

Or, should I write my application differently? I need to read out data from SPI periodically and preciously (4kHz => 250 us).

**Comment by cfriedt** [comment]
cc @tbursztyka

**Comment by owain-home** [comment]
I am also seeing this issue.

Does the SPI drivers being written this way mean no meaningful work can be done in an ISR for a SPI peripheral; as generally this means accessing registers via SPI bus. Must all such processing be farmed off to a worker thread where one can perform SPI transactions?

**Reply by GeorgeCGV** [reply]
Yes, offload threads are used for such cases.

**Comment by Oleh-Kravchenko** [comment]
Oh my gosh,
I've spent two days by debugging this...

Please implement some warning, or just make it working 😄
